src/sage/schemes/projective/cohomology.py

```diff
@@ -0,0 +1,500 @@
+            sage: c = X.structure_sheaf()._cohomology
+            sage: c.differential_top(1)
+            Vector space morphism represented as left-multiplication by the matrix:
+            []
+            Domain: Vector space of dimension 3 over Rational Field
+            Codomain: Vector space of dimension 0 over Rational Field
+        """
+        H1 = self.cohomology_group_top(i)
+        H0 = self.cohomology_group_top(i - 1)
+        M = self.resolution.differential(i).matrix()
+        K = self.coefficient_field
+        zero = K.zero()
+
+        assert M.ncols() == len(H1.summands_basis)
+        assert M.nrows() == len(H0.summands_basis)
+
+        A = []
+        for i in range(M.ncols()):
+            basis = H1.summands_basis[i]
+            for v in basis:
+                image = [zero for e in range(H0.rank)]
+                for j in range(M.nrows()):
+                    f = M[j,i]
+                    basis = H0.summands_basis[j]
+                    for c, m in zip(f.coefficients(), f.exponents()):
+                        u = v + vector(m)
+                        assert(sum(u) == -H0.shifts[j])
+                        if any(e >= 0 for e in u):
+                            continue
+                        k = H0.summands_index[j] + basis.index(u)
+                        image[k] += c
+                A.append(vector(K, image))
+
+        return H1.vector_space.hom(A, codomain=H0.vector_space, side='right')
+
+    def H(self, q):
+        r"""
+        Return the `q`-th cohomology group `H^q(\tilde M)`.
+
+        INPUT:
+
+        - ``q`` -- non-negative integer
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme([x^4 + y^4 + z^4])
+            sage: c = X.structure_sheaf()._cohomology
+            sage: c.H(1)
+            Vector space quotient V/W of dimension 3 over Rational Field where
+            V: Vector space of degree 3 and dimension 3 over Rational Field
+            Basis matrix:
+            [1 0 0]
+            [0 1 0]
+            [0 0 1]
+            W: Vector space of degree 3 and dimension 0 over Rational Field
+            Basis matrix:
+            []
+        """
+        r = self.projective_space_dimension
+        if q == r:
+            return self.cohomology_group_top(0).vector_space.quotient(self.differential_top(1).image())
+        elif 1 <= q and q < r:
+            return self.differential_top(r - q).kernel().quotient(self.differential_top(r - q + 1).image())
+        elif q == 0:
+            # 0th cohomology group of (global sections) is not realized in
+            # terms of cohomology groups and differential maps of twisted
+            # structure sheaves of the projective space.
+            return VectorSpace(self.coefficient_field, self.h(0))
+        elif q > r:
+            return VectorSpace(self.coefficient_field, 0)
+        raise IndexError('index must be non-negative')
```

Comment:
```suggestion
        return VectorSpace(self.coefficient_field, 0)
```
I think extending it by 0-dimensional spaces in both directions is the most natural.

Reply:
No. Cohomology group with negative index is not even defined. Cohomology groups do not form a complex.

Reply:
Ah, you may be misled because the class is called `MaruyamaComplex`. There are complexes behind, but the class itself is a structure to compute the cohomology groups from the complexes behind. Sorry for the confusing name.

Reply:
If needed, I may rename it perhaps simply to `Maruyama`.

Reply:
Okay, I kind of understand. I don't know the mathematics here and am taking guesses based on the PR. It would be probably be good to add a more precise description of what is being done (mathematically).

Technically the cohomology groups are still defined, just like for those extended for "large" indices, but they are just all trivial.

`Maruyama` by itself is not descriptive enough. Perhaps `MaruyamaComputation`?

Reply:
You claim that this is a complex, which is necessary to do the computation. Well, it's more of two complexes melded together from the description.

Reply:
As it is confusing, I renamed the class with `MaruyamaMethod`. It keeps the two complexes behind for computation purposes, but is not a mathematical object in itself.

Reply:
But it is a mathematical object; as you just said, it is a pair of complexes. However, I don’t see why the two complexes shouldn’t be separated.

Reply:
> But it is a mathematical object; as you just said, it is a pair of complexes.

A object of the class computes cohomology using the data about the pair of complexes. The cohomology, which is an integer, is not a property of the pair of complexes, it is a property of the sheaf.  It does not make a sense that the object "is" a pair of complexes. 

I wrote the class, and that is how I see it. Please respect my view.

> However, I don’t see why the two complexes shouldn’t be separated.

I don't understand. One complex is used to compute H(0), and the other is used to compute H(r) for r > 0.